src/main/java/org/folio/helper/CheckinHelper.java

```diff
@@ -82,29 +83,49 @@ private Future<ReceivingResults> processCheckInPieces(CheckinCollection checkinC
+            var dstTenantId = checkInPiece.getReceivingTenantId();
             if (checkInPiece.getId().equals(piece.getId()) && Boolean.TRUE.equals(checkInPiece.getCreateItem())) {
-              futures.add(purchaseOrderLineService.getOrderLineById(poLineId, requestContext)
+              pieceFutures.add(purchaseOrderLineService.getOrderLineById(poLineId, requestContext)
                 .map(PoLineCommonUtil::convertToCompositePoLine)
-                .compose(compPOL -> pieceCreateFlowInventoryManager.processInventory(compPOL, piece,
-                  checkInPiece.getCreateItem(), requestContext))
-                .map(aVoid -> Pair.of(poLineId, piece)));
+                .compose(compPol -> pieceCreateFlowInventoryManager.processInventory(compPol, piece, checkInPiece.getCreateItem(), requestContext))
+                .map(voidResult -> new PiecesHolder.PiecePoLineDto(poLineId, piece)));
+            } else if (checkInPiece.getId().equals(piece.getId()) && Objects.nonNull(srcTenantId) && Objects.nonNull(dstTenantId) && !srcTenantId.equals(dstTenantId)) {
+              pieceFutures.add(purchaseOrderLineService.getOrderLineById(poLineId, requestContext)
+                .map(PoLineCommonUtil::convertToCompositePoLine)
+                .map(compPol -> new PiecesHolder.PiecePoLineDto(compPol, piece, checkInPiece)));
             } else {
-              futures.add(Future.succeededFuture(Pair.of(poLineId, piece)));
+              pieceFutures.add(Future.succeededFuture(new PiecesHolder.PiecePoLineDto(poLineId, piece)));
             }
           })));
         return null;
       })
-      .compose(v -> collectResultsOnSuccess(futures).map(poLineIdVsPieceList -> StreamEx.of(poLineIdVsPieceList)
-        .distinct()
-        .groupingBy(Pair::getKey, mapping(Pair::getValue, collectingAndThen(toList(), lists -> StreamEx.of(lists)
-          .collect(toList()))))));
+      .compose(v1 -> collectResultsOnSuccess(pieceFutures)
+        .map(piecePoLineDtoList -> {
+          prepareItemsToRecreate(holder, piecePoLineDtoList);
+          return createPoLineIdPiecePair(piecePoLineDtoList);
+        }));
+  }
+
+  private static void prepareItemsToRecreate(PiecesHolder holder, List<PiecesHolder.PiecePoLineDto> piecePoLineDtoList) {
+    var itemRecreateDtoMap = new HashMap<String, PiecesHolder.PiecePoLineDto>();
+    piecePoLineDtoList.stream()
+      .filter(PiecesHolder.PiecePoLineDto::isRecreateItem)
+      .forEach(dto -> itemRecreateDtoMap.put(dto.getPoLineId(), dto));
+    holder.withItemsToRecreate(itemRecreateDtoMap);
+  }
+
+  private static Map<String, List<Piece>> createPoLineIdPiecePair(List<PiecesHolder.PiecePoLineDto> piecePoLineDtoList) {
+    return StreamEx.of(piecePoLineDtoList)
+      .map(v2 -> Pair.of(v2.getPoLineId(), v2.getPieceFromStorage())).distinct()
+      .groupingBy(Pair::getKey, mapping(Pair::getValue, collectingAndThen(toList(), lists -> StreamEx.of(lists).collect(toList()))));
```

Comment:
Could you clarify why the second parameter is used for `mapping()` method as `collectingAndThen(toList(), lists -> StreamEx.of(lists).collect(toList()))` and not `toList()` directly?

Reply:
> Could you clarify why the second parameter is used for `mapping()` method as `collectingAndThen(toList(), lists -> StreamEx.of(lists).collect(toList()))` and not `toList()` directly?

Yeah it's redundant, was there before, but I will improve it